File: flink-java-core/src/main/java/com/atguigu/flink/day4/KeyedProcessExamplePTime.java
package com.atguigu.flink.day4;

import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.util.Collector;

/**
 * @author ：hmz
 * @date ：Created in 2020/9/14 16:18
 */
public class KeyedProcessExamplePTime {

    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        env.socketTextStream("hadoop102", 9999)
                .map(new MapFunction<String, Tuple2<String, String>>() {
                    @Override
                    public Tuple2<String, String> map(String value) throws Exception {
                        String[] data = value.split(" ");
                        return Tuple2.of(data[0], data[1]);
                    }
                })
                .keyBy(r -> r.f0)
                .process(new MyKeyedProcess())
                .print();

        env.execute();
    }

    public static class MyKeyedProcess extends KeyedProcessFunction<String, Tuple2<String, String>, String> {

        @Override
        public void processElement(Tuple2<String, String> value, Context ctx, Collector<String> out) throws Exception {
            ctx.timerService().registerProcessingTimeTimer(ctx.timerService().currentProcessingTime() + 10 * 1000L);
        }

        @Override
        public void onTimer(long timestamp, OnTimerContext ctx, Collector<String> out) throws Exception {
            super.onTimer(timestamp, ctx, out);
            out.collect("定时器触发 time=" + timestamp);
        }
    }

}

File: flink-scala-core/src/main/scala/com/atguigu/flink/day4/KeyedProcessExamplePTime.scala
package com.atguigu.flink.day4

import org.apache.flink.streaming.api.functions.KeyedProcessFunction
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector

object KeyedProcessExamplePTime {

    def main(args: Array[String]): Unit = {
        val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
        env.setParallelism(1)

        env.socketTextStream("hadoop102", 9999, '\n')
                .map(line => {
                    val data: Array[String] = line.split(" ")
                    (data(0), data(1))
                })
                .keyBy(_._1)
                .process(new KeyedProcessFunction[String, (String, String), String] {
                    override def processElement(value: (String, String), ctx: KeyedProcessFunction[String, (String, String), String]#Context, out: Collector[String]) = {
                        ctx.timerService().registerProcessingTimeTimer(ctx.timerService().currentProcessingTime() + 10 * 1000L)
                        //println(ctx.timestamp() + "  --  " + ctx.timerService().currentProcessingTime())
                    }

                    override def onTimer(timestamp: Long, ctx: KeyedProcessFunction[String, (String, String), String]#OnTimerContext, out: Collector[String]): Unit = {
                        out.collect("触发定时器 time =" + timestamp)
                    }
                })
                .print()
        env.execute()
    }

}
